crates/pallet-executor/src/lib.rs

```diff
@@ -371,22 +376,30 @@ mod pallet {
                         );
                         return InvalidTransactionCode::Bundle.into();
                     }
-                    let primary_number = signed_opaque_bundle.bundle.receipt.primary_number;
+
+                    let first_primary_number = signed_opaque_bundle
+                        .bundle
+                        .receipts
+                        .get(0)
+                        .expect("Receipts in a bundle must be non-empty; qed")
+                        .primary_number;
 
                     let builder = ValidTransaction::with_tag_prefix("SubspaceSubmitBundle")
                         .priority(TransactionPriority::MAX)
-                        .and_provides(primary_number)
+                        .and_provides(first_primary_number)
```

Comment:
provides and requires should match each other, if you provide multiple numbers but require one number, they'll never match.

Reply:
I meant calling `and_provides` multiple times, so that it provides multiple things, my assumption was that later transactions can depend on any of those things.

Either way I believe I found another issue here: it is broken if there is more than one ER because it'll "provide" the first block primary block here, while next bundle's ER with "require" the last primary block, so they will never actually match each other.

Reply:
You can provide multiple things, but they are effectively the same because the receipt must be accepted consecutively, only the first primary block number matters and once it's included in some block, all the receipts will be stored on the primary chain.

>while next bundle's ER with "require" the last primary block, so they will never actually match each other.

Not sure I fully understand this. The next bundle's ER is always tied to the best receipt on the primary chain and unrelated to the receipts in some tx which is still in the tx pool.

Reply:
You have `and_requires` below that if submitted out of order will wait for `first_primary_number - One::one()` to arrive first. Now imagine that `first_primary_number - One::one()` is actually a second ER in the bundle. Since you only pick and provide the first primary number from the bundle, there may be no transaction that will provide the number expected.

Reply:
That makes sense, it's a valid edge case, and now I also agree with providing multiple receipt numbers is inherently more correct.